libkbfs/unflushed_path_cache.go

```diff
@@ -0,0 +1,328 @@
+}
+
+// appendToCache returns true when successful, and false if it needs
+// to be retried after the per-revision map is recomputed.
+func (upc *unflushedPathCache) appendToCache(irmd ImmutableRootMetadata,
+	perRevMap unflushedPathsPerRevMap) bool {
+	upc.lock.Lock()
+	defer upc.lock.Unlock()
+	switch upc.state {
+	case upcUninitialized:
+		// Nothing to do.
+	case upcInitializing:
+		// Append to queue for processing at the end of initialization.
+		upc.appendQueue = append(upc.appendQueue, irmd)
+	case upcInitialized:
+		if perRevMap == nil {
+			// This was prepared before `upc.chainsPopulator` was set,
+			// and needs to be done again.
+			return false
+		}
+		// Update the cache with the prepared paths.
+		upc.unflushedPaths[irmd.Revision()] = perRevMap
+	default:
+		panic(fmt.Sprintf("Unknown unflushedPathsCache state: %v", upc.state))
+	}
+	return true
+}
+
+func (upc *unflushedPathCache) removeFromCache(rev MetadataRevision) {
+	upc.lock.Lock()
+	defer upc.lock.Unlock()
+	switch upc.state {
+	case upcUninitialized:
+		// Nothing to do.
+	case upcInitializing:
+		// Append to queue for processing at the end of initialization.
+		upc.removeQueue = append(upc.removeQueue, rev)
+	case upcInitialized:
+		delete(upc.unflushedPaths, rev)
+	default:
+		panic(fmt.Sprintf("Unknown unflushedPathsCache state: %v", upc.state))
+	}
+}
+
+func (upc *unflushedPathCache) setCacheIfPossible(cache unflushedPathsMap,
+	cpp chainsPathPopulator) []ImmutableRootMetadata {
+	upc.lock.Lock()
+	defer upc.lock.Unlock()
+	if len(upc.appendQueue) > 0 {
+		// We need to process more appends!
+		queue := upc.appendQueue
+		upc.appendQueue = nil
+		return queue
+	}
+
+	for _, rev := range upc.removeQueue {
+		delete(cache, rev)
+	}
+	upc.removeQueue = nil
+	upc.unflushedPaths = cache
+	upc.chainsPopulator = cpp
+	close(upc.ready)
+	upc.ready = nil
+	upc.state = upcInitialized
+	return nil
+}
+
+// initialize should only be called when the caller saw a `true` value
+// from `startInitializeOrWait()`.  It returns the unflushed paths
+// associated with `irmds`.  If it returns a `false` boolean, the
+// caller must abort the initialization (although as long as `err` is
+// nil, the returns unflushed paths may be used).
+func (upc *unflushedPathCache) initialize(ctx context.Context,
+	uid keybase1.UID, kid keybase1.KID, codec kbfscodec.Codec,
+	log logger.Logger, cpp chainsPathPopulator,
+	irmds []ImmutableRootMetadata) (unflushedPathsMap, bool, error) {
+	// First get all the paths for the given range.  On the first try
+	unflushedPaths := make(unflushedPathsMap)
+	log.CDebugf(ctx, "Initializing unflushed path cache with %d revisions",
+		len(irmds))
+	err := addUnflushedPaths(ctx, uid, kid, codec, log, irmds, cpp,
+		unflushedPaths)
+	if err != nil {
+		return nil, false, err
+	}
+
+	initialUnflushedPaths := make(unflushedPathsMap)
+	// Only need to deep-copy the outer level of the map; the inner
+	// level (per-revision) is immutable.
```

Comment:
same comment as above, perhaps "is immutable" -> "shouldn't be modified by the caller"